dll/win32/shell32/CUserNotification.cpp

```diff
@@ -0,0 +1,637 @@
+    IN HICON hIcon,
+    IN LPCWSTR pszToolTip)
+{
+    NOTIFYICONDATAW nid = {0};
+
+    /* Destroy our local icon copy */
+    if (m_hIcon)
+        ::DestroyIcon(m_hIcon);
+
+    if (hIcon)
+    {
+        /* Copy the icon from the user */
+        m_hIcon = ::CopyIcon(hIcon);
+    }
+    else
+    {
+        /* Use the same icon as the one for the balloon if specified */
+        UINT uIcon = (m_dwInfoFlags & NIIF_ICON_MASK);
+        LPCWSTR pIcon = NULL;
+
+        if (uIcon == NIIF_INFO)
+            pIcon = IDI_INFORMATION;
+        else if (uIcon == NIIF_WARNING)
+            pIcon = IDI_WARNING;
+        else if (uIcon == NIIF_ERROR)
+            pIcon = IDI_ERROR;
+        else if (uIcon == NIIF_USER)
+            pIcon = NULL;
+
+        if (pIcon)
+            m_hIcon = ::LoadIconW(NULL, pIcon);
+        else
+            m_hIcon = NULL;
+    }
+
+    m_szTip = pszToolTip;
+
+    /* Update the notification icon (if we have one) */
+    if (!m_hWorkerWnd)
+        return S_OK;
+
+    /* Modify the notification icon - Ignore Shell_NotifyIconW() return value as on Windows */
+    SetUpNotifyData(NIF_ICON, &nid);
+    ::Shell_NotifyIconW(NIM_MODIFY, &nid);
+    return S_OK;
+}
+
+
+LRESULT CALLBACK
+CUserNotification::WorkerWndProc(
+    IN HWND hWnd,
+    IN UINT uMsg,
+    IN WPARAM wParam,
+    IN LPARAM lParam)
+{
+    /* Retrieve the current user notification object stored in the window extra bits */
+    CUserNotification* pThis = reinterpret_cast<CUserNotification*>(::GetWindowLongPtrW(hWnd, 0));
+    ASSERT(pThis);
+    ASSERT(hWnd == pThis->m_hWorkerWnd);
+
+    TRACE("Msg = 0x%x\n", uMsg);
+    switch (uMsg)
+    {
+        /*
+         * We do not receive any WM_(NC)CREATE message since worker windows
+         * are first created using the default window procedure DefWindowProcW.
+         * It's only subsequently that the window procedure is changed
+         * to the user one.
+         * We however receive WM_(NC)DESTROY messages.
+         */
+        case WM_NCDESTROY:
+        {
+            ::SetWindowLongPtrW(hWnd, 0, (LONG_PTR)NULL);
+            pThis->m_hWorkerWnd = NULL;
+            return 0;
+        }
+
+        case WM_TIMER:
+        {
+            TRACE("WM_TIMER(0x%lx)\n", wParam);
+
+            /* Destroy the associated timer */
+            ::KillTimer(hWnd, (UINT_PTR)wParam);
+
+            if (wParam == ID_BALLOON_TIMEOUT)
+            {
+                /* Timeout interval timer expired, it's time to display the balloon again */
+                NOTIFYICONDATAW nid = {0};
+
+                /* Modify the notification icon */
+                pThis->SetUpNotifyData(NIF_INFO, &nid);
+                ::Shell_NotifyIconW(NIM_MODIFY, &nid);
+            }
+            else if (wParam == ID_BALLOON_DELAYREMOVE)
+            {
+                /* Delay-remove timer expired, it's time to remove the notification */
+                pThis->RemoveIcon();
+                ::PostQuitMessage(0);
```

Comment:
If you are Calling the notify icon object from another window, does it post the quit message only to your own?

Reply:
To be checked, but since I pump in this current thread all the window messages using the message loop inside the Show() function, I should be able to detect the message first and handle it (aka. will finish destroying the window & quit the message loop & exit Show() function).

Reply:
I modified a bit this code, please check latest commits (where I use the flag `m_bIsShown`).